src/controls.h

```diff
@@ -2,11 +2,59 @@
 #define CONTROLS_H
 
 #include <sp2/io/keybinding.h>
+#include <cstdio>
+
+class KeyState
+{
+public:
+    KeyState(bool pressed, bool down, bool up, float value) : pressed(pressed), down(down), up(up), value(value) {}
+    KeyState() : pressed(false), down(false), up(false), value(0) {}
+
+    bool pressed;       //True when this key is currently being pressed.
+    bool down;          //True for 1 update cycle when the key is pressed.
+    bool up;            //True for 1 update cycle when the key is released.
+    float value;        //Returns a value in the range -1 to 1 for this keybinding. On keyboard keys this is always 0 or 1, but for joysticks this can be anywhere in the range -1.0 to 1.0
+
+    void writeToFile(FILE* f);
+
+    static bool readFromFile(FILE* f, KeyState& result);
+    static KeyState fromIO(const sp::io::Keybinding& key);
+};
+
+class PlayerControlsState
+{
+public:
+    KeyState up;
+    KeyState down;
+    KeyState left;
+    KeyState right;
+    KeyState primary_action;
+    KeyState secondary_action;
+    KeyState unknown2;
+    KeyState self_destruct;
+    KeyState unknown4;
+    KeyState unknown5;
+    KeyState start;
+
+    void writeToFile(FILE* f);
+
+    static bool readFromFile(FILE* f, PlayerControlsState& result);
```

Comment:
Why is this a static function?

Reply:
It doesn't need any information from an instance. It could be a non class member function but I find this a good place as well.

Reply:
It's an out parameter to work around the limitation of C++ that it's not easy to return multiple variables. Now I think of it, I could use the new C++ Optional template class. I can also rename the parameter to "out_result". It's up to you :)

Reply:
Now I see how you got there, wanting to make a static function that returns this object, but still a none static function:
```
bool readFromFile(FILE* f);
```
Would work exactly the same way, and matches the usage of writeToFile.

Also note, disk access on the Pi is very slow. So the write could potentially slow things down in the end...

Reply:
I find an non-static function not really appropriate because it doesn't use information from the instance it's called on. Let's agree to disagree. It's mostly a syntax difference. I'll change it :)